src/acl/DomainData.cc

```diff
@@ -78,24 +80,27 @@ aclDomainCompare(T const &a, T const &b)
             // When a.example.com comes after .example.com in an ACL
             // sub-domain is ignored. That is okay. Just important
             bool d3big = (strlen(d3) > strlen(d4)); // Always suggest removing the longer one.
-            debugs(28, DBG_IMPORTANT, "WARNING: '" << (d3big?d3:d4) << "' is a subdomain of '" << (d3big?d4:d3) << "'");
-            debugs(28, DBG_IMPORTANT, "WARNING: You should remove '" << (d3big?d3:d4) << "' from the ACL named '" << AclMatchedName << "'");
+            const auto longer = d3big ? d3 : d4;
+            const auto shorter = d3big ? d4 : d3;
+            debugs(28, DBG_IMPORTANT, "WARNING: '" << longer << "' is a subdomain of '" << shorter << "'" <<
+                    Debug::Extra << "You should remove '" << longer << "' from the ACL");
```

Comment:
It looks a little strange that we warn in the '.example.com a.example.com' case but throw in a reversed 'a.example.com .example.com' case. This difference was introduced at 14e563c, explaining that things are not fatal if a 'super-set wildcard' is kept. However if we consider dstdomain as a simple list of domains, the order should not matter - the fact that there are conflicting names should not depend on the position in the list. I would throw in both cases - and the code would also benefit from that.

Reply:
Thank you for this research. I bet the difference between the treatment of those two cases is not in the concepts behind them (you are right that they are conceptually the same from configuration point of view), but in _code_ required to preserve correct behavior: It is a lot more difficult to remove an existing entry (from inside of a comparison function in the middle of an insertion sequence!) than to avoid adding a new one...

> I would throw in both cases 

That would not be a backward compatible change. Moreover, there is nothing catastrophic in listing an address covered by another one (from configuration point of view, it is just a bit wasteful); ideally, we should _warn_ in both cases. If we are to solve this problem, we should be trying to move towards a warning rather than a catastrophic rejection.

> and the code would also benefit from that.

Yes, it is tempting to solve this problem first.

Now that we understand what is going on here (and have test cases to catch our mistakes), I wonder whether we can solve this problem without heroic efforts? Instead of trying to avoid problematic entries in the middle of tree insertion, ACLDomainData::parse() should:

* Use domains->find() to find an existing problematic entry (using a custom comparison function[^1]).
* If no old entry is found, insert the new entry (using a standard comparison function).
* If the found old entry is "longer", keep it. Warn about ignoring the new/shorter entry.
* If the found old entry is "shorter", remove it. Warn about ignoring the old/shorter entry. Insert the new entry (using a standard comparison function).

[^1]: That function should return 0 (i.e. "match") for test cases 3, 4, and 5. It should not print any warnings or errors in those cases -- the code will be easier to grok if ACLDomainData::parse() will take care of that after getting a match from find().

I am sure we can make the above "work". The only question I have is whether we can prove or demonstrate that the above algorithm does not double the tree construction time. With STL, we would be able to give the insertion code "hints" about the new entry position, avoiding repeating the search. I hope that Splay tree auto-tuning properties result in essentially the same effect even though Splay::find() does not return a position and Splay::insert() does not accept hints (and Splay::delete() cannot use a position).

Can you demonstrate those expected Splay tree effects by hacking Splay tree to count the number of SPLAYCMP calls (i.e. the insertion "cost") and then measuring the cost of insertion of an entry that

1. mismatches all existing N entries and is bigger than the N-th stored entry
2. mismatches all existing N entries and is smaller than the n-th stored entry
3. matches a n-th stored entry exactly
4. matches a n-th stored entry and is "smaller" than that entry
5. matches a n-th stored entry and is "bigger" than that entry

... where N is, say, 3 (and n is iterated from 1 to N). AFAICT, the above should result in 1+(5-1)*3=13 test cases (with hard-coded temporary loops to measure the above).

I hope that we can show that the cost counters in _each_ of the 13 test cases do not differ much when comparing old/official reject-while-inserting code and the new find+insert implementation. AFAICT, both implementations should show identical counters for the first two cases. Counters for find+insert will be slightly bigger for cases 3-5 because of Splay::delete() cost.

Reply:
> Deltas for N=30: 3, 3, 0, -, 0

My Test1 was incorrect because it tried to insert a host (xxx.example.com) which was less than some of the existing hosts (e.g, ay.example.com > xxx.example.com). I fixed it and here are the correct deltas, which do not depend on N: 1, 3, 0, -, 0. 
Also I adjusted the test to check all host permutations in the source array (to cover all possible splay tree combinations). I tested for small N (<6), the results are the same.
Also I tested the old algorithm (in Test4) which would throw in the official code:
```
Test4: matches a n-th stored entry and is smaller than that entry
N: 3
3/7(3f + 2r + 2fi)
3/12(3f + 3r + 6fi)
1/6(1f + 3r + 2fi)
Delta(abs,max): 9; Delta(%,mean): 411.111%
```
For other N the increase is ~ 300%.

Reply:
> How can there be 6 find+insert comparisons

There are initially 3 elements [ab.ab.example.com, ab.ab.example.com, ac.ac.example.com] and we try to add a new ".ab.example.com". After it finds and removes the second element "ab.ab.example.com", the three has "aa.aa.example.com" at its head and a leaf "ac.ac.example.com". 

The "find" operation performs 3 comparisons in "SplayNode<V>::splay()":
1. "aa.aa.example.com" < ".ab.example.com" (top=="aa.aa.example.com")
2. "ac.ac.example.com" > ".ab.example.com" (top=="aa.aa.example.com")
3. "aa.aa.example.com" < ".ab.example.com" (top=="ac.ac.example.com")

The "insert" operation also performs  3 comparisons in "SplayNode<V>::splay()":
1. "ac.ac.example.com" > ".ab.example.com" (top=="ac.ac.example.com")
2. "aa.aa.example.com" < ".ab.example.com" (top=="ac.ac.example.com")
3. "ac.ac.example.com" > ".ab.example.com" (top=="aa.aa.example.com")

The "extra" comparison occurs because the "for(;;)" loop in splay() updates the "top" variable before the next iteration and the same value is checked twice (as at 1 and 3 steps above). However the difference is that the node at the (3) step does not have leaves (but it had at (1)) step - so the loop ends.

Reply:
> here are the correct deltas, which do not depend on N: 1, 3, 0, -, 0.

> I tested for small N (<6), the results are the same.

> Test4 for N=3: Delta: 411%. For other N the increase is ~ 300%.

Just to double check, that "other N" statement talks about N values like 1, 2, 4, and 5, right? And, for all those N values, the deltas do not steadily grow with N (i.e. they stay constant after reaching some maximum value), correct? In other words, the new code introduces some overhead in a few corner cases, but that overhead is bounded by a small constant value (e.g., 10 extra comparisons or less,  even when N is, say, one million).

Reply:
> that "other N" statement talks about N values like 1, 2, 4, and 5, right?

IIRC I checked for some N>3, including big values (say, 30-40 for the test without permutations). These tests showed that  Delta does not grow with N, but is more or less about a constant 300% (depending on the N and permutation). I think this result can be explained by the fact that mean values for all three coefficients in sums like (3/12(3f + 3r + 6fi)) are equal.